### ISSUE TITLE
Proposal: Bindings as custom attributes (ko-*)

### ISSUE THREAD

**Issue by robertsundstrom** [knockout/knockout#1806]
*This proposal has been edited.*

The use of data-bindings as well as declarative ways of defining them has really become a trend in modern front-end framework design.

Knockout was a pioneer with that, enabling developers to use data-binding, through observables, and to do it declaratively using the *data-bind* attribute.

However, another trend that has come up with frameworks (both client-side and server-side) such as Angular, Aurelia and ASP.NET is to use custom prefixed attributes, or *"directives"* as they are called in Angular. For example: ng-model.

This has proven to make intents clearer and easier to understand as the code also gets cleaner.

This proposal is about considering to implement something similar in Knockout, *as an alternative to the existing way*.

## Reason ##

Simply put: It is cleaner and easier to understand than the *data-bind* syntax.

It could also boost Knockouts popularity as an alternative to big frameworks like Angular and Aurelia.

## Syntax ##

The proposed feature would make it possible to declare bindings as HTML-attributes prefixed with a namespace. For the default bindings it would be *ko-*, similar tor *ng-* in angular.

To demonstrate the difference between the current way of declaring bindings and the proposed way, consider this trivial example found at Knockout's website:

    <div>
        You've clicked <span data-bind="text: numberOfClicks"></span> times
        <button data-bind="click: incrementClickCounter">Click me</button>
    </div>

This could with the new proposed way of declaring bindings be expressed like so:

    <div>
        You've clicked <span ko-text="numberOfClicks"></span> times
        <button ko-click="incrementClickCounter">Click me</button>
    </div>

Here *ko-* is being used as a prefix, similar to *ng-* and *aspnet-* etc.

Some examples showing what attributes would exist:

* ko-visible
* ko-text
* ko-css
* ko-click
* ko-value

Consider a custom binding in namespace "go":

* go-fade

etc...

## Implementation ##

This can be implemented as a plug-in. *(edit)*

Bindings, whether inbuilt or custom, would simply map to *ko-<binding>*. 

Once read by Knockout they would be processed similarly to how it is being done today with *data-bind*.

## Related feature: String interpolation ##

There are other things to consider: for example, adding string interpolation syntax that enables the developer to write bindings directly in the HTML.

It could be based on the proposed ES6 syntax, as currently used by Aurelia:

````text 
<h2>Hello ${name}!</h2>
````

A separate proposal needs to be elaborated upon.

## Personal notes ##

~It is unclear to me what this would take to be implemented as I have just barely started to look into the source code for Knockout.~

It has, since the time of first writing, become clear to me that you can build your own providers.

**Comment by stalniy** [comment]
@robertsundstrom this can be implemented in a separate library/plugin (separate bindingProvider). And on other hand there are at least 2 library which do what you suggested:
- @mbest knockout.punches - https://github.com/mbest/knockout.punches
- and mine ko-mustashed - https://github.com/stalniy/ko-mustached

**Comment by robertsundstrom** [comment]
Thank you, @stalniy. I will look into it.

**Comment by robertsundstrom** [comment]
I started implementing my own binding provider in my own skeleton SPA repo.

https://github.com/robertsundstrom/knockoutjs-express-es6

There is a lot of room for improvements.

**Comment by frialex** [comment]
Though ko.punches improves upon an old ko pain point, it also introduces a problem:
What is written in an editor is not the same as what is displayed in the elements panel of dev-tools.

The transformation from mustache syntax to knockout binding syntax makes debugging difficult since a mental juggling is placed upon the developer.

+1 for ko- prefixed binding syntax!

**Comment by IanYates** [comment]
This is intriguing.  I agree with the op's remarks about the similarity to Angular, etc which may appeal to some.  I have a new dev on my team and he's familiar with Angular but not with KO.  He's got the hang of data-bind now (and we use KO punches mustashed syntax too) but the attribute-based approach would help with that transition and also keep the code clean.
This thread encouraged me to really read the KO punches documentation again, line by line, as I knew it could do so much more but hadn't taken the time to really absorb it.  Frankly it's really well designed (no surprises there given the author) - I particularly like the namespace handling where style.color: [[stuffHere]] can be bound and its transparently treated as style: {colour: [[stuffHere]]}.  The attribute approach here though would be the icing on the cake :)
:+1:

**Comment by robertsundstrom** [comment]
As I mentioned, I wrote my own binding provider. I also started to rewrite it in order to allow custom prefixes/namespaces. That has not been pushed yet.

**Comment by claudeduguay** [comment]
I'd like to mention that the way custom binding management is implemented in Knockout's current version (3.3.0) does not permit this approach to be used in Knockout components, which I think would have increased value as it would be nice to follow the same model as the Web Component specifications. The libraries/extensions mentioned above predate the component feature do not address this at all.

I've tried on two occasions to write a custom binding manager and got tangled up in having to re-implement incremental portions of the underlying implementations in order to get to the component-related code. I finally gave up (during both excursions) as too much of the code was inaccessible via standard APIs and copying the underlying code is really bad practice and was too involved to be practical. 

To put that another way, I believe, based on my own two attempts, that the component feature does not allow custom attribute bindings through the binding manager API.

I'd love to see an example of how to do this. In fact, I strongly believe that if one of the Knockout team members tried to do this, it would flush out the underlying API problems I mentioned and either lead to fixes in future versions of Knockout or to an illustration that other could use and proof are no significant obstacles actually exist 'if you know what you're doing'. ;-)

Requirement: A simple custom binding manager example that lets you bind element attributes on Knockout components (something with a ko- prefix would be completely appropriate).

**Comment by brianmhunt** [comment]
@claudeduguay You're absolutely right about the underlying API. The API part will hopefully be addressed when #1369 gets addressed.

**Comment by claudeduguay** [comment]
Thanks Brian. 

Looks like you pointed me to to a Closure issue though (why?).

What's the expected timeline for 3.4 (roughly, not asking for a commitment here, just curious)? Is it nearing critical-mass in terms of additions and/or fixes you guys want to get into the next version?

**Comment by brianmhunt** [comment]
@claudeduguay Sorry for not being clear :) The closure issue relates to the privacy of internal versus external API items. In particular the closure compiler minifies anything not explicitly exported, making them inaccessible. Switching to an uglify-compatible minification will still probably end up exporting those internal methods.

Not sure about 3.4 offhand.

**Comment by robertsundstrom** [comment]
Anyway. We see how other frameworks jump lightyears ahead with their first-class support for ES6. I do, however, not think about breaking compatibility, even if that might be logical in the future.

And not to forget to mention, the Knockout website feels outdated and incomplete in some sense and that deters some people who find it confusing. There needs to be better samples.

If we want to keep Knockout relevant and attractive then there needs to be some action from the community.

**Comment by tarwn** [comment]
TLDR Version: +1 to making it possible as a plugin, -1 to the belief that this style is cleaner, -100 to making it the default for knockout

I don't find the separate attributes style more readable. Currently it reads like a chunk of JSON assigned to a single binding attribute, which has helped when learning/teaching some of the attributes that accept objects (consider what a template or CSS attribute would look like) and when pointing out the difference for certain bindings between passing an observable versus passing a literal with observables in it. It also feels more naturally to assume a list of properties in a single attribute is going to be applied in order, versus assuming that a list of attributes on an element won't be modified in any way by the server or client and will be operated on in order (for example, the issue with virtual bindings and browsers adding elements in ways that can break them). 

Compatibility is very important for me, and limiting breaking changes between version is very important. My company builds a SaaS product that is never finished, we are constantly adding features, refactoring major bits, evolving it forward as we and the market become more sophisticated. We'll take on the occasional cost of testing/rewriting small bits for breaking changes, but the Angular and ASP.Net "you just have to rewrite your app" approach of breaking changes actually makes those tools not cost effective for us. I'm not saying that anyone has suggested knockout change how it handles that, but it felt like an important point to bring up as we compare to Angular (which I personally don't find to be a shining example on the documentation, patterns, readability, getting started fronts, version compatibility, or browser compatibility fronts).

So final thought, I absolutely agree that building a plugin that offers this capability should be a possibility and that hopefully the fixes above should bring about that ability for components. But I don't believe that changing all of knockout over to this style as the default is _proven_ to be any cleaner or any better than the current style and it would actively be detrimental to those of us that have chosen it for larger apps (I know, define "larger," I'm thinking big enough to be a long multi-person rewrite, so ~100kloc+?).

**Comment by robertsundstrom** [comment]
The strength of Knockout is it minimalistic but yet extensible core. The library does of course come with its own default binding provider and component loader etc. but you, as a developer, can replace it if you want. That shouldn't change. That is an important part of what knockout is and it should be stressed.

But we need better documentation for all the features (in particular the bindings) and better plug-in discoverability.

**Comment by claudeduguay** [comment]
I don't think Knockout should be refactored to accommodate alternative tag styles. What it needs to do is make the existing binding manager extension point work for both the more mature binding handler model AND the newer component model without an implementer having to resort to major surgery, otherwise it's not meeting the intent of the extension point. This got broken when component handling was added and simply needs to be fixed.

Knockout cannot and should not try to be all things to all people. So long as a simple extension can be implemented to achieve the results you are looking for, you are not restricted in expressing that desire or in publishing a simple library that meets more specific needs.

I would add, however, that 'enabling' alignment  with the standard Web Component model is highly desirable, even if there is only a mention on the site (and not necessarily adding it directly to the library), preferably with an example of how a developer could write a suitable binding manager. In practice, if the APIs are clean, the binding manager should apply equally to both standard binding handlers and components transparently.

In passing, I find it a little confusing to group this item into a bucket with an apparently unrelated issue as I don't think suitable methods will automatically be exposed in the aforementioned fix. At the moment the existing extension point it's not fully compliant and it is not possible to use the binding manager to modify the behavior of component (params) bindings. This seems like an oversight that should be addressed to bring the library back into alignment.

**Comment by mbest** [comment]
I'm closing this since I don't think we should include this syntax in Knockout natively, and it's possible to do this already with plugins.
